plugins/flytekit-greatexpectations/flytekitplugins/great_expectations/task.py

```diff
@@ -0,0 +1,256 @@
+import datetime
+import logging
+import os
+import shutil
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Type, Union
+
+import great_expectations as ge
+from dataclasses_json import dataclass_json
+from great_expectations.checkpoint import SimpleCheckpoint
+from great_expectations.core.run_identifier import RunIdentifier
+from great_expectations.core.util import convert_to_json_serializable
+from great_expectations.exceptions import ValidationError
+
+from flytekit import PythonInstanceTask
+from flytekit.core.context_manager import FlyteContext
+from flytekit.extend import Interface
+from flytekit.types.file.file import FlyteFile
+from flytekit.types.schema import FlyteSchema
+
+
+@dataclass_json
+@dataclass
+class BatchRequestConfig(object):
+    """
+    Use this configuration to configure Batch Request. A BatchRequest can either be
+    a simple BatchRequest or a RuntimeBatchRequest.
+
+    Args:
+        data_connector_query: query to request data batch
+        runtime_parameters: parameters to be passed at runtime
+        batch_identifiers: identifiers to identify the data batch
+        batch_spec_passthrough: reader method if your file doesn't have an extension
+    """
+
+    data_connector_query: Optional[Dict[str, Any]] = None
+    runtime_parameters: Optional[Dict[str, Any]] = None
+    batch_identifiers: Optional[Dict[str, str]] = None
+    batch_spec_passthrough: Optional[Dict[str, Any]] = None
+
+
+class GreatExpectationsTask(PythonInstanceTask[BatchRequestConfig]):
+    """
+    This task can be used to validate your data.
+
+    Args:
+        name: name of the task
+        datasource_name: tell where your data lives and how to get it
+        expectation_suite_name: suite which consists of the data expectations
+        data_connector_name: connector to identify data batches
+        inputs: inputs to pass to the execute() method
+        is_runtime: do you want to create data batches at runtime?
```

Comment:
Hi @samhita-alla - I have mixed feelings about this `is_runtime` argument. I interpreted the main purpose of the `BatchRequestConfig` class as a funnel for taking in both RuntimeBatchRequests and BatchRequests, so that the user wouldn't need to think about this. But if we are requiring the user to specify `is_runtime` anyway, I wonder if it makes sense for the user to pass in a typed GreatExpectations `BatchRequest` or `RuntimeBatchRequest`, and then just test which one it is. I think that might make the interface more straightforward. What do you think?

Reply:
Hi, @talagluck. Even now, `BatchRequestConfig` class is a funnel for taking in both `RuntimeBatchRequest` and `BatchRequest`. However, since it isn't possible to determine if the request is run-time or not automatically, we need the user's input. If `is_runtime` is set to `True`, the plugin would consider the user-given dataset as the `batch_data` or the `query`. The user needn't give one separately in such cases. Let me explain the use-case, w.r.t Flyte.

In Flyte, we enclose all our code in tasks and workflows. Great Expectations in Flyte can be a task or a type. If we have to feed input, we do that through the `inputs` argument in the **task** (line 51 in task.py), and the dataset in itself is considered the input in **type**. Say, as you suggested, we use Great Expectations' builtin `RuntimeBatchRequest`, the user then has to pass input (`batch_data` or `query`), which cannot be done this way. It can only be initialized and sent from outside, which wouldn't be within the scope of Flyte's entities, and henceforth, wouldn't respect Flyte's data flow. 

For example, this is how it's being done now wherein the user needn't pay attention to setting the runtime parameter:

```python
runtime_task_obj = GreatExpectationsTask(
    name="greatexpectations.task.runtime",
    datasource_name="my_pandas_datasource",
    inputs=kwtypes(dataframe=FlyteSchema),
    expectation_suite_name="test.demo",
    data_connector_name="my_runtime_data_connector",
    is_runtime=True,
    task_config=BatchRequestConfig(
        batch_identifiers={
            "pipeline_stage": "validation",
        },
    ),
    context_root_dir=CONTEXT_ROOT_DIR,
)

@workflow
def runtime_wf(dataset: str = DATASET_LOCAL) -> None:
    dataframe = ...(csv_file=dataset)
    runtime_task_obj(dataframe=dataframe)
```

The plugin sets the `dataframe` as `batch_data`, which otherwise needs to be initialized outside of the workflow and set within.

Reply:
Thank you for this explanation, @samhita-alla - I think I get it now. I see now that at the time the user defines the `runtime_task_obj`, they are not ready to create the BatchRequest, so they are creating a sort of pre-BatchRequest that is completed and instantiated when the task is actually run.

My only suggestion here would be to consider parsing the Great Expectations config to determine the type of the `data_connector` instead of requiring the specification of an additional argument. As a very rough example, you could do something like this:

```python
specified_datasource = "my_pandas_datasource"
specified_data_connector = "my_runtime_data_connector"

[my_datasource] = [i for i in context.list_datasources() if i.get("name")==specified_datasource]

data_connector_class_lookup = {data_connector_name: data_connector_class.get("class_name") for data_connector_name, data_connector_class in my_datasource[0].get("data_connectors").items()}

specified_data_connector_class = data_connector_class_lookup.get(specified_data_connector)

if specified_data_connector_class = "RuntimeDataConnector":
...

```

Reply:
Thank you! Made the required changes.